diff --git a/chromium/build/common.gypi b/chromium/build/common.gypi
index 272a42f..12a586d 100644
--- a/chromium/build/common.gypi
+++ b/chromium/build/common.gypi
@@ -2665,6 +2665,7 @@
 
         # QtWebEngine: To support other clang versions:
         '-Wno-tautological-compare',
+        '-Wno-empty-body',
         '-Wno-unknown-attributes',
         '-Wno-unknown-warning-option',
       ],
@@ -3670,7 +3671,8 @@
           'libraries_for_target%': '',
           'conditions' : [
             # Enable -Wextra for chromium_code when we control the compiler.
-            ['clang==1', { 'wextra': '-Wextra' }, { 'wextra': '-Wno-extra' }],
+            # But not for QtWebEngine were we do not.
+            ['clang==1 and use_qt==0', { 'wextra': '-Wextra' }, { 'wextra': '-Wno-extra' }],
           ],
         },
         'defines': [
@@ -5919,8 +5921,8 @@
                     'AdditionalOptions': [
                       '-fsanitize=address',
                       '-fsanitize-blacklist=<(PRODUCT_DIR)/../../tools/memory/asan/blacklist_win.txt',
-		      # Omit variable info to speed up /Z7 links.
-		      '-gline-tables-only',
+                      # Omit variable info to speed up /Z7 links.
+                      '-gline-tables-only',
                     ],
                     'AdditionalIncludeDirectories': [
                       # MSVC needs to be able to find the sanitizer headers when
diff --git a/chromium/content/content_child.gypi b/chromium/content/content_child.gypi
index 5dff9ca..16d89241 100644
--- a/chromium/content/content_child.gypi
+++ b/chromium/content/content_child.gypi
@@ -6,7 +6,6 @@
     '../base/base.gyp:base',
     '../components/mime_util/mime_util.gyp:mime_util',
     '../components/scheduler/scheduler.gyp:scheduler',
-    '../components/tracing.gyp:tracing',
     '../components/webcrypto/webcrypto.gyp:webcrypto',
     '../ipc/ipc.gyp:ipc',
     '../mojo/mojo_base.gyp:mojo_common_lib',
diff --git a/chromium/content/content_common.gypi b/chromium/content/content_common.gypi
index 72e182b..730f6fb 100644
--- a/chromium/content/content_common.gypi
+++ b/chromium/content/content_common.gypi
@@ -7,7 +7,6 @@
     '<(DEPTH)/third_party/khronos/khronos.gyp:khronos_headers',
     '../base/base.gyp:base',
     '../cc/cc.gyp:cc',
-    '../components/tracing.gyp:tracing',
     '../device/bluetooth/bluetooth.gyp:device_bluetooth',
     '../gpu/blink/gpu_blink.gyp:gpu_blink',
     '../gpu/command_buffer/command_buffer.gyp:gles2_utils',
