diff --git a/libavcodec/cdxl.c b/libavcodec/cdxl.c
index 7e7b7bc..8c9e400 100644
--- a/libavcodec/cdxl.c
+++ b/libavcodec/cdxl.c
@@ -19,8 +19,6 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 #include "libavutil/intreadwrite.h"
 #include "libavutil/imgutils.h"
 #include "avcodec.h"
diff --git a/libavcodec/h263dec.c b/libavcodec/h263dec.c
index 405d3c6..6e3d885 100644
--- a/libavcodec/h263dec.c
+++ b/libavcodec/h263dec.c
@@ -25,8 +25,6 @@
  * H.263 decoder.
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 #include "libavutil/cpu.h"
 #include "internal.h"
 #include "avcodec.h"
diff --git a/libavcodec/h264.c b/libavcodec/h264.c
index 502ee92..a1e0585 100644
--- a/libavcodec/h264.c
+++ b/libavcodec/h264.c
@@ -25,8 +25,6 @@
  * @author Michael Niedermayer <michaelni@gmx.at>
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 #include "libavutil/imgutils.h"
 #include "libavutil/opt.h"
 #include "internal.h"
diff --git a/libavcodec/h264_cabac.c b/libavcodec/h264_cabac.c
index 60048e0..28d2a15 100644
--- a/libavcodec/h264_cabac.c
+++ b/libavcodec/h264_cabac.c
@@ -26,7 +26,6 @@
  */
 
 #define CABAC(h) 1
-#define UNCHECKED_BITSTREAM_READER 1
 
 #include "config.h"
 #include "cabac.h"
diff --git a/libavcodec/h264_cavlc.c b/libavcodec/h264_cavlc.c
index dad819f..2c23cae 100644
--- a/libavcodec/h264_cavlc.c
+++ b/libavcodec/h264_cavlc.c
@@ -26,7 +26,6 @@
  */
 
 #define CABAC(h) 0
-#define UNCHECKED_BITSTREAM_READER 1
 
 #include "internal.h"
 #include "avcodec.h"
diff --git a/libavcodec/h264_parser.c b/libavcodec/h264_parser.c
index 3b5b652..f373a1c 100644
--- a/libavcodec/h264_parser.c
+++ b/libavcodec/h264_parser.c
@@ -25,8 +25,6 @@
  * @author Michael Niedermayer <michaelni@gmx.at>
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 #include "parser.h"
 #include "h264data.h"
 #include "golomb.h"
diff --git a/libavcodec/ituh263dec.c b/libavcodec/ituh263dec.c
index 5d604ef..a28ed52 100644
--- a/libavcodec/ituh263dec.c
+++ b/libavcodec/ituh263dec.c
@@ -27,8 +27,6 @@
  * h263 decoder.
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 //#define DEBUG
 #include <limits.h>
 
diff --git a/libavcodec/mpeg12.c b/libavcodec/mpeg12.c
index 1099e17..ecb67ed 100644
--- a/libavcodec/mpeg12.c
+++ b/libavcodec/mpeg12.c
@@ -25,8 +25,6 @@
  * MPEG-1/2 decoder
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 //#define DEBUG
 #include "libavutil/avassert.h"
 #include "libavutil/timecode.h"
diff --git a/libavcodec/mpeg4video_parser.c b/libavcodec/mpeg4video_parser.c
index 3cbd69d..e085669 100644
--- a/libavcodec/mpeg4video_parser.c
+++ b/libavcodec/mpeg4video_parser.c
@@ -20,8 +20,6 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 #include "parser.h"
 #include "mpegvideo.h"
 #include "mpeg4video.h"
diff --git a/libavcodec/mpeg4videodec.c b/libavcodec/mpeg4videodec.c
index be4aa37..342707b 100644
--- a/libavcodec/mpeg4videodec.c
+++ b/libavcodec/mpeg4videodec.c
@@ -20,8 +20,6 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
  */
 
-#define UNCHECKED_BITSTREAM_READER 1
-
 #include "libavutil/opt.h"
 #include "error_resilience.h"
 #include "internal.h"
