--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_el.ts	2017-08-31 10:40:34.146005995 -0700
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;ΑΚΥΡΟ</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ΑΚΥΡΟ</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ΑΚΥΡΟ</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ΑΚΥΡΟ</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ΑΚΥΡΟ</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_nb.ts	2017-08-31 10:43:18.770822325 -0700
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Avbryt</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_nl.ts	2017-08-31 10:44:58.663317665 -0700
@@ -714,7 +714,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/dbustray/qdbustrayicon.cpp" line="+293"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Oké</translation>
     </message>
 </context>
 <context>
@@ -754,7 +754,7 @@
         <location filename="../../qtbase/src/widgets/dialogs/qmessagebox.cpp" line="+1972"/>
         <location line="+446"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Oké</translation>
     </message>
 </context>
 <context>
@@ -984,7 +984,7 @@
     <message>
         <location line="+1"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Oké</translation>
     </message>
 </context>
 <context>
@@ -1833,7 +1833,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp" line="+779"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Oké</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Annuleren</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4236,7 +4236,7 @@
     <message>
         <location filename="../../qtbase/src/gui/kernel/qplatformtheme.cpp" line="+698"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Oké</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Annuleren</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4499,7 +4499,7 @@
     <message>
         <location line="+1"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Oké</translation>
     </message>
     <message>
         <location filename="../../qtbase/src/printsupport/kernel/qplatformprintdevice.cpp" line="+308"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Annuleren</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Annuleren</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Annuleren</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_pt.ts	2017-08-31 10:46:08.195662458 -0700
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Cancelar</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Cancelar</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Cancelar</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Cancelar</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Cancelar</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_ro.ts	2017-08-31 10:47:04.463941477 -0700
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Anulează</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Anulează</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Anulează</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Anulează</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Anulează</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_sv.ts	2017-08-31 10:47:48.432159504 -0700
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Avbryt</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>Avbryt</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_th.ts	2017-08-31 10:49:18.044603868 -0700
@@ -714,7 +714,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/dbustray/qdbustrayicon.cpp" line="+293"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>ตกลง</translation>
     </message>
 </context>
 <context>
@@ -754,7 +754,7 @@
         <location filename="../../qtbase/src/widgets/dialogs/qmessagebox.cpp" line="+1972"/>
         <location line="+446"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>ตกลง</translation>
     </message>
 </context>
 <context>
@@ -984,7 +984,7 @@
     <message>
         <location line="+1"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;ตกลง</translation>
     </message>
 </context>
 <context>
@@ -1833,7 +1833,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp" line="+779"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;ตกลง</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;ยกเลิก</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4236,7 +4236,7 @@
     <message>
         <location filename="../../qtbase/src/gui/kernel/qplatformtheme.cpp" line="+698"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>ตกลง</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ยกเลิก</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4499,7 +4499,7 @@
     <message>
         <location line="+1"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>ตกลง</translation>
     </message>
     <message>
         <location filename="../../qtbase/src/printsupport/kernel/qplatformprintdevice.cpp" line="+308"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ยกเลิก</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ยกเลิก</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>ยกเลิก</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_tr.ts	2017-08-31 10:50:42.845024370 -0700
@@ -714,7 +714,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/dbustray/qdbustrayicon.cpp" line="+293"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Tamam</translation>
     </message>
 </context>
 <context>
@@ -754,7 +754,7 @@
         <location filename="../../qtbase/src/widgets/dialogs/qmessagebox.cpp" line="+1972"/>
         <location line="+446"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Tamam</translation>
     </message>
 </context>
 <context>
@@ -984,7 +984,7 @@
     <message>
         <location line="+1"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Tamam</translation>
     </message>
 </context>
 <context>
@@ -1833,7 +1833,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp" line="+779"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;Tamam</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;İptal</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4236,7 +4236,7 @@
     <message>
         <location filename="../../qtbase/src/gui/kernel/qplatformtheme.cpp" line="+698"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Tamam</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>İptal</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4499,7 +4499,7 @@
     <message>
         <location line="+1"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>Tamam</translation>
     </message>
     <message>
         <location filename="../../qtbase/src/printsupport/kernel/qplatformprintdevice.cpp" line="+308"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>İptal</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>İptal</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>İptal</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_zh_CN.ts	2017-08-31 10:52:18.257497495 -0700
@@ -714,7 +714,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/dbustray/qdbustrayicon.cpp" line="+293"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>确定</translation>
     </message>
 </context>
 <context>
@@ -754,7 +754,7 @@
         <location filename="../../qtbase/src/widgets/dialogs/qmessagebox.cpp" line="+1972"/>
         <location line="+446"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>确定</translation>
     </message>
 </context>
 <context>
@@ -984,7 +984,7 @@
     <message>
         <location line="+1"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;确定</translation>
     </message>
 </context>
 <context>
@@ -1833,7 +1833,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp" line="+779"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;确定</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;取消</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4236,7 +4236,7 @@
     <message>
         <location filename="../../qtbase/src/gui/kernel/qplatformtheme.cpp" line="+698"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>确定</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4499,7 +4499,7 @@
     <message>
         <location line="+1"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>确定</translation>
     </message>
     <message>
         <location filename="../../qtbase/src/printsupport/kernel/qplatformprintdevice.cpp" line="+308"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
     <message>
         <location line="+2"/>
--- qtbase_untranslated.ts	2017-08-31 09:10:42.507270302 -0700
+++ qtbase_zh_TW.ts	2017-08-31 10:55:20.710402230 -0700
@@ -714,7 +714,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/dbustray/qdbustrayicon.cpp" line="+293"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>確定</translation>
     </message>
 </context>
 <context>
@@ -754,7 +754,7 @@
         <location filename="../../qtbase/src/widgets/dialogs/qmessagebox.cpp" line="+1972"/>
         <location line="+446"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>確定</translation>
     </message>
 </context>
 <context>
@@ -984,7 +984,7 @@
     <message>
         <location line="+1"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;確定</translation>
     </message>
 </context>
 <context>
@@ -1833,7 +1833,7 @@
     <message>
         <location filename="../../qtbase/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp" line="+779"/>
         <source>&amp;OK</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;確定</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -1843,7 +1843,7 @@
     <message>
         <location line="+2"/>
         <source>&amp;Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>&amp;取消</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4236,7 +4236,7 @@
     <message>
         <location filename="../../qtbase/src/gui/kernel/qplatformtheme.cpp" line="+698"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>確定</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4296,7 +4296,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
     <message>
         <location line="+2"/>
@@ -4499,7 +4499,7 @@
     <message>
         <location line="+1"/>
         <source>OK</source>
-        <translation type="unfinished"></translation>
+        <translation>確定</translation>
     </message>
     <message>
         <location filename="../../qtbase/src/printsupport/kernel/qplatformprintdevice.cpp" line="+308"/>
@@ -4853,7 +4853,7 @@
     <message>
         <location filename="../../qtbase/src/widgets/dialogs/qprogressdialog.cpp" line="+188"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
 </context>
 <context>
@@ -7149,7 +7149,7 @@
     <message>
         <location line="+4"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
     <message>
         <location line="+1"/>
@@ -8476,7 +8476,7 @@
     <message>
         <location line="+2"/>
         <source>Cancel</source>
-        <translation type="unfinished"></translation>
+        <translation>取消</translation>
     </message>
     <message>
         <location line="+2"/>
