diff --git a/third_party/libtiff/tiffconf.h b/third_party/libtiff/tiffconf.h
index bcc0bc5..ca749da 100644
--- a/third_party/libtiff/tiffconf.h
+++ b/third_party/libtiff/tiffconf.h
@@ -39,7 +39,11 @@
 //fx_system.h already include the string.h in ANSIC
 
 /* Define to 1 if you have the <search.h> header file. */
-/*#define HAVE_SEARCH_H 1 */
+#if _FXM_PLATFORM_ == _FXM_PLATFORM_WINDOWS_ && _MSC_VER >= 1900
+// search.h is always available in VS 2015 and above, and may be
+// available in earlier versions.
+#define HAVE_SEARCH_H 1
+#endif
 
 /* The size of a `int', as computed by sizeof. */
 /* According typedef int	int32_t; in the fx_system.h*/
