diff --git a/javascript/atoms/device.js b/javascript/atoms/device.js
index e2693f8..3731db3 100644
--- a/javascript/atoms/device.js
+++ b/javascript/atoms/device.js
@@ -627,6 +627,7 @@ bot.Device.prototype.maybeToggleOption = function() {
   // Only WebKit fires the change event itself and only for multi-selects,
   // except for Android versions >= 4.0.
   if (!(goog.userAgent.WEBKIT && select.multiple) ||
+      (goog.userAgent.product.CHROME && bot.userAgent.isProductVersion(28)) ||
       (goog.userAgent.product.ANDROID && bot.userAgent.isProductVersion(4))) {
     bot.events.fire(select, bot.events.EventType.CHANGE);
   }
diff --git a/javascript/chrome-driver/build.desc b/javascript/chrome-driver/build.desc
index 423c450..1889095 100644
--- a/javascript/chrome-driver/build.desc
+++ b/javascript/chrome-driver/build.desc
@@ -24,6 +24,11 @@ js_fragment(name = "get_page_zoom",
   function = "webdriver.chrome.getPageZoom",
   deps = [ ":deps" ])
 
+js_fragment(name = "touch_single_tap",
+  module = "bot.action",
+  function = "bot.action.tap",
+  deps = [ "//javascript/atoms:deps" ])
+
 js_test(name = "test",
   srcs = [ "test/*_test.html" ],
   deps = [
@@ -39,6 +44,7 @@ js_fragment_cpp(name = "atoms",
     "//javascript/chrome-driver:get_location_in_view:chrome",
     "//javascript/chrome-driver:get_page_zoom:chrome",
     "//javascript/chrome-driver:is_element_clickable:chrome",
+    "//javascript/chrome-driver:touch_single_tap:chrome",
     "//javascript/webdriver/atoms:clear:chrome",
     "//javascript/webdriver/atoms:clear_local_storage:chrome",
     "//javascript/webdriver/atoms:clear_session_storage:chrome",
