diff --git a/third_party/tlslite/tlslite/constants.py b/third_party/tlslite/tlslite/constants.py
index ceaa903..52c20ac 100644
--- a/third_party/tlslite/tlslite/constants.py
+++ b/third_party/tlslite/tlslite/constants.py
@@ -224,7 +224,7 @@ class CipherSuite:
     def getSrpCertSuites(settings):
         return CipherSuite._filterSuites(CipherSuite.srpCertSuites, settings)
 
-    srpAllSuites = srpSuites + srpCertSuites
+    srpAllSuites = srpCertSuites + srpSuites
 
     @staticmethod
     def getSrpAllSuites(settings):
