diff -pu a/nss/lib/ssl/ssl3ecc.c b/nss/lib/ssl/ssl3ecc.c
--- a/nss/lib/ssl/ssl3ecc.c	2014-01-03 19:28:03.550814608 -0800
+++ b/nss/lib/ssl/ssl3ecc.c	2014-01-03 19:40:46.523288747 -0800
@@ -1073,6 +1073,7 @@ static const PRUint8 ecPtFmt[6] = {
 static PRBool
 ssl3_SuiteBOnly(sslSocket *ss)
 {
+#if 0
     /* See if we can support small curves (like 163). If not, assume we can
      * only support Suite-B curves (P-256, P-384, P-521). */
     PK11SlotInfo *slot =
@@ -1086,6 +1087,9 @@ ssl3_SuiteBOnly(sslSocket *ss)
     /* we can, presume we can do all curves */
     PK11_FreeSlot(slot);
     return PR_FALSE;
+#else
+    return PR_TRUE;
+#endif
 }
 
 /* Send our "canned" (precompiled) Supported Elliptic Curves extension,
