--- source/common/putil.c
+++ source/common/putil.c
@@ -37,15 +37,17 @@
 ******************************************************************************
 */
 
-/* Define _XOPEN_SOURCE for Solaris and friends. */
-/* NetBSD needs it to be >= 4 */
-#if !defined(_XOPEN_SOURCE)
-#if __STDC_VERSION__ >= 199901L
-/* It is invalid to compile an XPG3, XPG4, XPG4v2 or XPG5 application using c99 on Solaris */
-#define _XOPEN_SOURCE 600
+/* Define _XOPEN_SOURCE for access to POSIX functions. */
+#ifdef _XOPEN_SOURCE
+    /* Use the predefined value. */
 #else
-#define _XOPEN_SOURCE 4
-#endif
+    /*
+     * Version 6.0:
+     * The Open Group Base Specifications Issue 6 (IEEE Std 1003.1, 2004 Edition)
+     * also known as
+     * SUSv3 = Open Group Single UNIX Specification, Version 3 (UNIX03)
+     */
+#   define _XOPEN_SOURCE 600
 #endif
 
 /* Make sure things like readlink and such functions work.
