diff -ur assimpGIT/code/ColladaExporter.cpp assimp/code/ColladaExporter.cpp
--- assimpGIT/code/ColladaExporter.cpp	2016-12-08 14:13:07.000000000 -0500
+++ assimp/code/ColladaExporter.cpp	2016-12-08 09:23:22.000000000 -0500
@@ -383,6 +383,7 @@
         case aiLightSource_SPOT:
             WriteSpotLight(light);
             break;
+        case aiLightSource_AREA:
         case aiLightSource_UNDEFINED:
         case _aiLightSource_Force32Bit:
             break;
