diff -ur assimpGIT/contrib/unzip/crypt.h assimp/contrib/unzip/crypt.h
--- assimpGIT/contrib/unzip/crypt.h	2016-12-07 20:20:54.000000000 -0500
+++ assimp/contrib/unzip/crypt.h	2016-12-07 17:00:44.000000000 -0500
@@ -51,7 +51,7 @@
     (*(pkeys+1)) += (*(pkeys+0)) & 0xff;
     (*(pkeys+1)) = (*(pkeys+1)) * 134775813L + 1;
     {
-      register int keyshift = (int)((*(pkeys+1)) >> 24);
+      int keyshift = (int)((*(pkeys+1)) >> 24);
       (*(pkeys+2)) = CRC32((*(pkeys+2)), keyshift);
     }
     return c;
