diff --git a/third_party/freetype2/include/ftconfig.h b/third_party/freetype2/include/ftconfig.h
index 9ea3546..16f8854 100644
--- a/third_party/freetype2/include/ftconfig.h
+++ b/third_party/freetype2/include/ftconfig.h
@@ -397,9 +397,9 @@ FT_BEGIN_HEADER
 #ifndef FT_EXPORT
 
 #ifdef __cplusplus
-#define FT_EXPORT( x )  extern "C"  x
+#define FT_EXPORT( x )  __attribute__((visibility("default"))) extern "C"  x
 #else
-#define FT_EXPORT( x )  extern  x
+#define FT_EXPORT( x )  __attribute__((visibility("default"))) extern  x
 #endif
 
 #endif /* !FT_EXPORT */
@@ -408,9 +408,9 @@ FT_BEGIN_HEADER
 #ifndef FT_EXPORT_DEF
 
 #ifdef __cplusplus
-#define FT_EXPORT_DEF( x )  extern "C"  x
+#define FT_EXPORT_DEF( x )  __attribute__((visibility("default"))) extern "C"  x
 #else
-#define FT_EXPORT_DEF( x )  extern  x
+#define FT_EXPORT_DEF( x )  __attribute__((visibility("default"))) extern  x
 #endif
 
 #endif /* !FT_EXPORT_DEF */
