diff --git a/data/Makefile.am b/data/Makefile.am
index 9af4ca1..7df2cf5 100644
--- a/data/Makefile.am
+++ b/data/Makefile.am
@@ -2,13 +2,16 @@
 # Install, and make if needed, prboom.wad
 #
 
-noinst_PROGRAMS = rdatawad
-rdatawad_SOURCES = rd_main.c
-rdatawad_SOURCES += rd_util.c rd_util.h
-rdatawad_SOURCES += rd_output.c rd_output.h
-rdatawad_SOURCES += rd_sound.c rd_sound.h
-rdatawad_SOURCES += rd_palette.c rd_palette.h
-rdatawad_SOURCES += rd_graphic.c rd_graphic.h
+#noinst_PROGRAMS = rdatawad
+#rdatawad_SOURCES = rd_main.c
+#rdatawad_SOURCES += rd_util.c rd_util.h
+#rdatawad_SOURCES += rd_output.c rd_output.h
+#rdatawad_SOURCES += rd_sound.c rd_sound.h
+#rdatawad_SOURCES += rd_palette.c rd_palette.h
+#rdatawad_SOURCES += rd_graphic.c rd_graphic.h
+
+rdatawad: rd_main.c rd_util.c rd_output.c rd_sound.c rd_palette.c rd_graphic.c
+	gcc -o $@ $^ -I..
 
 waddir=$(DOOMWADDIR)
 wad_DATA=prboom-plus.wad
